lib/core/objects.toit

```diff
@@ -187,24 +187,91 @@ class LazyInitializer_:
     // The __invoke_initializer__ builtin does a tail call to the method with the given id.
     return __invoke_initializer__ id_or_task_
 
+  initializing -> Task_:
+    if id_or_task_ is Task_: return id_or_task_
+    if id_or_task_ is InitializingTask_: return (id_or_task_ as InitializingTask_).task
+    throw "BAD_LAZY_INITIALIZER_CALL"
+
+  add_waiting task/Task_:
+    if id_or_task_ is Task_: id_or_task_ = InitializingTask_ id_or_task_
+    if not id_or_task_ is InitializingTask_: throw "BAD_LAZY_INITIALIZER_CALL"
+    (id_or_task_ as InitializingTask_).add_waiting task
+
+  do_waiting [block]:
+    if id_or_task_ is Task_: return
+    if not id_or_task_ is InitializingTask_: throw "BAD_LAZY_INITIALIZER_CALL"
+    current := (id_or_task_ as InitializingTask_).waiting
+    while current:
+      block.call current.task
+      current = current.next
+
+
+class InitializingTask_:
```

Comment:
I would have written this without the two new Task wrappers. Tasks can be chained together themselves while blocked (they have a field for that) and I'd be okay searching for the current task in a chained list to throw the recursive in-progress exception.

So the task in the LazyInitializer becomes the head of a linked list of tasks (the last one isn't blocked before evaluating the initializer). Not sure I care about the order of waking these up, so I'd probably just do it LIFO.

Reply:
Tried that and it didn't work. (Spent quite some time to understand why).
The initializing task can run more code, and in particular, it can run code that involves mutexes, ...
Thinking even more, the currently uploaded PR doesn't work either if a task has already another task in the `blocked` list...
Will revert to the initial approach.